README.md
# GirlCrawler

![example](https://raw.githubusercontent.com/Ericlong233/girlcrawler/master/res/example.png)

[Github](https://github.com/Ericlong233/girlcrawler) | [npm](https://www.npmjs.com/package/girlcrawler) | [更新日志](https://github.com/Ericlong233/girlcrawler/blob/master/CHANGELOG.md) | [欢迎 Star](https://github.com/Ericlong233/girlcrawler)

*注意：适当撸妹益脑，过度撸妹伤身。请在遵守当地法律法规的前提下使用。煎蛋网资源有限，请适当使用。*

### 介绍

GirlCrawler 是一个简洁、高效的[煎蛋妹子图](http://jandan.net/ooxx)爬虫。它基于高效、异步的 Node.js。具有以下特性：

* 快速，由 Node.js 的异步 I/O 提供支持
* 基于 OO/XX 的过滤器
* 友好的 CLI 界面
* 通过 MD5 自动过滤失效图片

### 用法

`girlcrawler [options]`

* `-t, --thread <thread>` 下载*最大*并发数，默认为 64。由于网速的快慢，即使这个值非常大，也可能不会有任何速度上的提升。
* `-f, --filter <filter>` 基于 OO/XX 的过滤器，默认为 `"oo >= 2 * xx"`。
* `-b, --buffer <buffer>` 缓冲，表示在下次更新时，从往前 N 页开始爬起。默认为 5。

### 安装

* **在线安装 (推荐)：** 执行`npm install girlcrawler -g` 。
* **手动安装：**
  1. 执行 `git clone http://github.com/Ericlong233/girlcrawler.git` 将项目克隆到本地。
  2. 执行 `npm install ./girlcrawler -g` 。

* 以后，可以直接执行 `girlcrawler` 来运行爬虫。图片保存在当前目录的 `jandangirls` 内。
* **注意：** 一定不要漏掉 `-g` ！需要卸载时，输入 `npm uninstall girlcrawler -g` 即可。

### 未来特性

*注意：此栏里的功能尚未实现，且随时有可能变化*

- [ ] ~~寻找图片的更大版本 (基于 Tineye 或 Google Images)~~ 没有合适的解决方案

### 开源组件

* [async](https://github.com/caolan/async) - 异步操作整理
* [commander](https://github.com/tj/commander.js/) - 命令行参数
* [request](https://github.com/request/request) - 向服务器发出请求
* [cheerio](https://github.com/cheeriojs/cheerio) - 解析 HTML
* [colors](https://github.com/Marak/colors.js) - 终端字体颜色
* [progress](https://github.com/visionmedia/node-progress) - 终端进度条
* [md5](https://github.com/pvorb/node-md5) - MD5 计算

### 协议

[MIT License](https://github.com/Ericlong233/girlcrawler/blob/master/LICENSE)


CHANGELOG.md
# GirlCrawler 更新日志

### 1.3.5

* 将 http://jandan.net 换成 http://i.jandan.net，减少对煎蛋流量的使用
  * 网页源码大小从 80~100KB 减少为 40~60KB

### 1.3.0

* 添加 `-b, --buffer` ，表示在下次更新时，从往前 N 页开始爬起。默认为 5
* 默认过滤器由 `"oo > xx"` 改为 `"oo >= 2 * xx"` ，提高图片质量

### 1.2.1

* 修复了一个 bug，有小几率使 md5 函数接收到 undefined，发生错误

### 1.2.0

* 识别图片 MD5，自动清除失效图片 (由 [md5](https://github.com/pvorb/node-md5) 模块提供支持)

### 1.1.5

* 修复了无法爬大图的 bug

### 1.1.4

* 修复了一个 bug，发生在 [3483878](http://jandan.net/ooxx/page-123#comment-3483878) (无图片回复块)

### 1.1.1

* 将 config.json 改为 .config
* 默认下载线程从 128 修改为 64

### 1.1.0

* 在每一次下载前写入配置，避免了下载中断无法继续进行的问题

### 1.0.0

* 初始版本 (未发布)


main.js
#!/usr/bin/env node

//
// Ericlong233/girlcrawler
//

var async    = require("async"),
    program  = require("commander"),
    request  = require("request"),
    cheerio  = require("cheerio"),
    colors   = require("colors"),
    Progress = require("progress"),
    md5      = require("md5"),
    fs       = require("fs"),
    path     = require("path");

var url    = "http://i.jandan.net/ooxx",
    dir    = "./jandangirls",
    config = dir + "/.config";

var DEFAULT_THREAD  = 64,
    DEFAULT_FILTER  = "oo >= 2 * xx",
    DEFAULT_BUFFER  = 5
    BREAK_IMAGE_MD5 = "9a49736345f17e6c90dfe3bcd74dfb5e";    // 这个不是下载下来的坏图的 md5, 是 request 爬下来的 body 的 md5

program.version("1.3.5")
       .description("简洁、高效的煎蛋妹子图爬虫, powered by Node.js")
       .option("-t, --thread <thread>", "下载" + "最大".underline + "并发数, 默认为 " + DEFAULT_THREAD)
       .option("-f, --filter <filter>", "基于 OO/XX 的过滤器, 默认为 \"" + DEFAULT_FILTER + "\"")
       .option("-b, --buffer <buffer>", "缓冲, 表示在下次更新时, 从往前 N 页开始爬起. 默认为 " + DEFAULT_BUFFER)
       .parse(process.argv);

function getPageCount(callback) {
    request(url, (err, response, body) => {
        var $ = cheerio.load(body);
        var pageCount = +$("span.current-comment-page").eq(0).text().replace(/(\[|\])/g, "");
        callback(null, pageCount);
    });
}

function getConfig(pageCount, callback) {
    var thread, filter, buffer, start, end = pageCount;
    if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir);
    }
    if (!fs.existsSync(config)) {
        thread = program.thread || DEFAULT_THREAD;
        filter = program.filter || DEFAULT_FILTER;
        buffer = program.buffer || DEFAULT_BUFFER;
        start  = 1;
    } else {
        var data = JSON.parse(fs.readFileSync(config, "utf8"));
        thread = program.thread || data.thread || DEFAULT_THREAD;
        filter = program.filter || data.filter || DEFAULT_FILTER;
        buffer = program.buffer || data.buffer || DEFAULT_BUFFER;
        if (data.lastUpdate)
            if (data.lastUpdate - buffer > 0) start = data.lastUpdate - buffer
            else start = 1
        else start = 1;
    }
    callback(null, thread, filter, buffer, start, end);
}

function setConfig(thread, filter, buffer, i, callback) {
    fs.writeFileSync(config, JSON.stringify({
        thread: thread, filter: filter, buffer: buffer, lastUpdate: i
    }));
    callback(null);
}

function getPictureInfos(page, filter, callback) {
    request(url + "/page-" + page + "#comments", (err, response, body) => {

        var $ = cheerio.load(body);
        async.waterfall([
            (callback) => {
                var comments = [];
                $("ol.commentlist li:not(.row)").each((i, element) => {
                    comments.push(cheerio.load($(element).html()));
                });
                callback(null, comments);
            },
            (comments, callback) => {
                var pictures = [];
                for (var item of comments) {
                    var id, urls = [], oo, xx;
                    oo = +item("span.tucao-unlike-container a.like + span").text();
                    xx = +item("span.tucao-unlike-container a.unlike + span").text();
                    if (!eval(filter)) continue;
                    id = item("span.tucao-unlike-container a.like").attr("data-id");
                    item("a.view_img_link").each((i, element) => {
                        urls.push("http:" + $(element).attr("href"));
                    });
                    if (urls.length > 1) {
                        var j = 1;
                        for (var i of urls) {
                            pictures.push({id: id + "-" + j++, url: i, oo: oo, xx: xx});
                        }
                    } else {
                        pictures.push({id: id, url: urls[0], oo: oo, xx: xx});
                    }
                }
                callback(null, pictures);
            },
        ], (err, result) => {
            callback(err, result);
        });

    });
}

function downloadPics(page, pictures, thread, callback) {

    var breakImage = []
    var bar = new Progress("Downloading page " + page + " [:bar]".blue + " :percent ", {
        total: pictures.length,
        width: 25,
        complete: "#",
        incomplete: "-",
        clear: true
    });

    async.mapLimit(pictures, thread, (item, callback) => {
        if (!item.url) return callback();    // fixed a bug in 3483878, page 123
        request({
            url: item.url,
            headers: {
                "Connection":    "keep-alive",
                "Cache-Control": "max-age=0",
                "User-Agent":    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.90 Safari/537.36"
            }
        }, (err, response, body) => {
            bar.tick();
            if (!body || (md5(body) == BREAK_IMAGE_MD5)) {
                breakImage.push(item.id + path.extname(item.url));
            }
            callback();
        }).pipe(fs.createWriteStream(dir + "/" + item.id + path.extname(item.url)));
    }, (err, result) => {
        for (var i of breakImage) {
            fs.unlinkSync(dir + "/" + i);
        }
        callback();
    })

}

// 虽然用了 async 还是有一些 callback hell...
async.waterfall([

    (callback) => getPageCount(callback),
    (pageCount, callback) => getConfig(pageCount, callback),
    (thread, filter, buffer, start, end, callback) => {

        console.log();
        console.log("爬虫运行中... 参数: " + ("thread=" + thread + ", filter=\"" + filter + "\", buffer=" + buffer).cyan.underline);

        // 遍历每一页
        var i = start;
        async.whilst(() => i <= end, (callback) => {
            async.waterfall([
                (callback) => setConfig(thread, filter, buffer, i, callback),
                (callback) => getPictureInfos(i, filter, callback),
                (pictures, callback) => downloadPics(i, pictures, thread, callback),
            ], (err, result) => {
                i++;
                callback(err, result);
            });
        }, (err, result) => {
            callback(err, result);
        });

    },

], (err, result) => {
    console.log("** F I N I S H E D ! **\n".rainbow)
});
